### PR TITLE
Added exception handler for request in the Middleware class of django.

### PR DESCRIPTION
**Pull request by snowmaksym**
Fixes: #70

### CODE REVIEW

**Review by apotterri — CHANGES_REQUESTED** [review]
It's hard to tell whether this actually addresses #70. Please add at least one test that shows that if `get_response` raises, the generated appmap matches an expected appmap.

File: appmap/django.py [hunk 1]
```diff
@@ -143,8 +150,42 @@ def __call__(self, request):
...
+                parent_id=self.call_event_id,
+                elapsed=duration,
+                exc_info=sys.exc_info()
+            )
+            self.recorder.add_event(exception_event)
+
+            self.call_event_id = None
+            self.start = 0
+
+            return HttpResponse("Error processing the request.", status=500)
```
Comment (dividedmind) [inline appmap/django.py:186]: What if the app wants to return a different response for the exception?

**Review by dividedmind — CHANGES_REQUESTED** [review]
Like Alan said, this needs a test. Also, wouldn't it be easier to simply add a `try` in the middleware instead? 

BTW, it's not clear to me when an exception might be raised in `get_response` — surely if the application code raises an exception, it will have been handled and turned into the correct response before it gets to us? @apotterri, do you remember if there was a specific scenario where it lead to broken appmaps, or was the intention to just be defensive?

**Comment by apotterri** [comment]
I agree with @dividedmind: I think it would be better to wrap `get_response` in a `try` block. It should create the `Exception` event and re-raise the exception.

An unhandled exception in a view will propagate through `get_response`. If we only ever mapped bug-free applications, we would be able to rely on their exception handling. :) It seems better to be defensive, and ensure that a correctly formed appmap is always generate.

**Comment by snowmaksym** [comment]
I've deleted `process_exception` function in `Middleware` class, and wrapped  `get_response` in a try block.
But I'm not sure if there is an exception when `get_response` function is called.

File: appmap/django.py [hunk 2]
```diff
@@ -131,7 +132,18 @@ def __call__(self, request):
             )
             self.recorder.add_event(call_event)
 
-        response = self.get_response(request)
+        try:
+            response = self.get_response(request)
+        except Exception as e:
```
Comment (dividedmind) [inline appmap/django.py:137]: You're not using `e`; might as well not bind it and just reraise with plain `raise`.

**Review by dividedmind — CHANGES_REQUESTED** [review]
This looks better now. Still needs a test though.

**Comment by apotterri** [comment]
> I've deleted `process_exception` function in `Middleware` class, and wrapped `get_response` in a try block.
> But I'm not sure if there is an exception when `get_response` function is called.

The way to be sure is to arrange for an exception to get raised from a view. To get started, I'd suggest adding a new route and its corresponding view to `test_django.py`. Then, add a test that requests the new route and verifies that the appropriate events are generated.

As a follow on task, we should move the new test into `web_framework.py` and add the appropriate function to `test_flask.py`.

File: appmap/test/test_django.py [hunk 3]
```diff
@@ -107,6 +111,7 @@ def echo_view(request):
     django.urls.path('post/<username>/<int:post_id>/summary', user_post_view),
     django.urls.re_path(r'^post/unnamed/(\d+)$', post_unnamed_view),
     django.urls.path('echo', echo_view),
+    django.urls.re_path('exception', exception_view)
```
Comment (apotterri) [inline appmap/test/test_django.py:114]: This can just be `path`, rather than `re_path`.

File: appmap/test/test_django.py [hunk 4]
```diff
@@ -116,6 +121,21 @@ def test_unnamed(client, events):
     # message attribute.
     assert len(events[0].message) == 0
 
+def test_exception(client, events):
+    try:
```
Comment (apotterri) [inline appmap/test/test_django.py:125]: You should use `pytest.raises` to verify that the expected exception is being raised: https://docs.pytest.org/en/reorganize-docs/new-docs/user/pytest_raises.html .

File: appmap/test/test_django.py [hunk 5]
```diff
@@ -116,6 +121,21 @@ def test_unnamed(client, events):
...
+        client.get('/exception')
+
+        assert events[0].http_server_request.items() >= {
+            'request_method': 'GET',
+            'path_info': '/exception',
+            'protocol': 'HTTP/1.1'
+        }.items()
+    except django.core.exceptions.PermissionDenied:
+        assert events[1].parent_id == events[0].id
+        assert events[1].exceptions[0].items() >= {
```
Comment (apotterri) [inline appmap/test/test_django.py:135]: To make this check look like others in the file, you should write it as:

```python
assert events[1].items() >= {
   'parent_id': events[0].id,
    'exceptions': [{
        'class': 'django.core.exceptions.PermissionDenied'
    ]}
}.items()
```

To set the `message` attribute of `exceptions[0]`, you need to pass a message to the exception constructor. For example, if you wrote:

```python
def exception_view(_request):
    raise RuntimeError('An error')
```

`exceptions[0].message` should be `'An Error'`.

**Review by apotterri — CHANGES_REQUESTED** [review]

File: appmap/test/test_django.py [hunk 6]
```diff
@@ -122,19 +121,24 @@ def test_unnamed(client, events):
     assert len(events[0].message) == 0
 
 def test_exception(client, events):
-    try:
+    with pytest.raises(RuntimeError):
         client.get('/exception')
 
+        # The 1st event: HttpServerRequestEvent
```
Comment (apotterri) [inline appmap/test/test_django.py:127]: These checks should be move out of the `with` block. Nothing except the request should raise a `RuntimeError`.

**Review by apotterri — CHANGES_REQUESTED** [review]

**Comment by apotterri** [comment]
Please add an entry to CHANGELOG for this fix as well.

File: appmap/test/test_django.py [hunk 7]
```diff
@@ -116,6 +120,26 @@ def test_unnamed(client, events):
...
+        client.get('/exception')
+
+        # The 2nd event: ExceptionEvent
+        assert events[1].items() >= {
+            'parent_id': events[0].id,
+            'exceptions': [{
+                'class': 'RuntimeError',
+                'message': 'An error'
+            }]
+        }.items()
```
Comment (dividedmind) [inline appmap/test/test_django.py:134]: This could be moved out of the `with` block.

**Review by dividedmind — CHANGES_REQUESTED** [review]
Looks good. But please use `git rebase --interactive` to squash the commits into one. No need for the commit history to reflect the development now that it's ready to merge. Also please try to use the conventional form for a commit message:

```
Change how widget operates (imperative, <60cols, no period)

This changes the widget to operate much better. This can be a 
whole proper paragraph (up to 80 cols) describing it in more detail.
You can also:
- point out various things of note here as desired,
- use bullet points and other light formatting,
- remember to make the subject succinct so it's informative,
  representative but short, '<verb> <detail>' form of what the commit does,
  suitable to understanding at a glance from `git log --graph --oneline`
  and similar commands.
```

File: appmap/test/test_django.py (hunk 7 above)
Comment (snowmaksym) [reply appmap/test/test_django.py:134]: We can't check `ExceptionEvent` if the `assert` block is moved out of the `with` block.

Comment (dividedmind) [reply appmap/test/test_django.py:134]: I don't see why not. Have you tried it?

Comment (snowmaksym) [reply appmap/test/test_django.py:134]: yeah, I've tried it, but it raises `AttributeError`.
`AttributeError: 'HttpServerResponseEvent' object has no attribute 'items'`

Comment (dividedmind) [reply appmap/test/test_django.py:134]: Ah, right, it does. But when it's in the `with` block the code just gets skipped, because `client.get` raises.

Comment (snowmaksym) [reply appmap/test/test_django.py:134]: Then I think the `HttpServerResponseEvent` should also inherit the `Exception` event to check the exception from a request.

Comment (dividedmind) [reply appmap/test/test_django.py:134]: No. And anyway, that's beside the point. Your test just doesn't do what it's supposed to. Go ahead, remove your changes in `appmap.django`. The test still passes.

**Comment by dividedmind** [comment]
@apotterri I'm not sure about these changes. I thought the whole point of #70 was to fix the issue of `get_response` raising an exception? If django catches an exception and responds with a 500, I think that should get recorded in an appmap, not an exception it has caught, even if it provides a hook for that. Also, won't using that hook potentially change the observable behavior of an app?

**Comment by snowmaksym** [comment]
@apotterri, @dividedmind 
If we don't need to catch the exception from a view in the middleware, we can record the response status code 500 and its response.
Let me know I have to change the `django.py` to record the 500 response instead of an exception.

**Comment by dividedmind** [comment]
> Let me know I have to change the `django.py` to record the 500 response instead of an exception.

It has already been doing this.

**Comment by snowmaksym** [comment]
But I think it's better to catch the exception rather than the 500 response, because it is more helpful to understand the dynamic structure and flow of the django app.

**Comment by dividedmind** [comment]
> But I think it's better to catch the exception rather than the 500 response, because it is more helpful to understand the dynamic structure and flow of the django app.

If an exception is raised in instrumented code, it will already be captured in the appmap.

**Comment by apotterri** [comment]
> @apotterri I'm not sure about these changes. I thought the whole point of #70 was to fix the issue of `get_response` raising an exception? If django catches an exception and responds with a 500, I think that should get recorded in an appmap, not an exception it has caught, even if it provides a hook for that. Also, won't using that hook potentially change the observable behavior of an app?

Yeah, you're right. Let me try again.

**Review by dividedmind — APPROVED** [review]
Looks ok now, thanks! Please remember to fix up the commits though before merging; it seems you gave the wrong reference to `--fixup`. Also, looking at the full diff, you'll notice you don't need to qualify `django.urls.path` anymore.

File: appmap/test/test_django.py [hunk 8]
```diff
@@ -150,6 +155,27 @@ def test_included_view(client, events):
...
+        'request_method': 'GET',
+        'path_info': '/exception',
+        'protocol': 'HTTP/1.1'
+    }.items()
+
+    assert events[1].parent_id == events[0].id
+    assert events[1].exceptions[0].items() >= {
+            'class': 'builtins.RuntimeError',
+            'message': 'An error'
+    }.items()
```
Comment (dividedmind) [inline appmap/test/test_django.py:177]: You can use DictIncluding now:
```suggestion
    assert events[0].http_server_request == DictIncluding({
        'request_method': 'GET',
        'path_info': '/exception',
        'protocol': 'HTTP/1.1'
    })

    assert events[1].parent_id == events[0].id
    assert events[1].exceptions == [DictIncluding({
            'class': 'builtins.RuntimeError',
            'message': 'An error'
    })]
```

**Comment by apotterri** [comment]
> Looks ok now, thanks! Please remember to fix up the commits though before merging; it seems you gave the wrong reference to `--fixup`. Also, looking at the full diff, you'll notice you don't need to qualify `django.urls.path` anymore.

Yeah, will do. I did a rebase with `--autosquash` (because I wanted you to see a distinct fixup commit), maybe that confused things.
